concordium-consensus/src/Concordium/GlobalState/LMDB/Helpers.hs

```diff
@@ -28,6 +30,158 @@ import Foreign.Marshal.Utils
 import Foreign.Marshal.Alloc
 import Foreign.Ptr
 import Foreign.Storable
+import Lens.Micro.Platform
+
+
+-- Implementation of reader-writer locks that strongly prefer writers.
+
+data RWState =
+   -- |Nobody has acquired the lock.
+  Free { waitingWriters :: !Word64 }
+  -- |The given number of readers have acquired the lock.
+  | ReadLocked {
+      readers :: !Word64,
+      waitingWriters :: !Word64
+      }
+  -- |The lock is acquired by a single writer.
+  | WriteLocked {
+      waitingWriters :: !Word64
+      }
+    deriving(Show)
+
+-- |A reader writer lock with a bias for writers. If there are waiting writers
+-- readers will block.
+data RWLock = RWLock {
+  state :: !(MVar RWState),
+  -- |Used to signal waiting readers to wake up. This is empty when there is at
+  -- least one active reader and full otherwise.
+  readLock :: !(MVar ()),
+  -- |Used to signal waiting writers to wake up. This is empty when there is an
+  -- active writer, and full otherwise.
+  writeLock :: !(MVar ())
+  }
+
+-- |Initialize a lock in the unlocked state.
+initializeLock :: IO RWLock
+initializeLock = do
+  state <- newMVar (Free 0)
+  readLock <- newMVar ()
+  writeLock <- newMVar ()
+  return RWLock{..}
+
+-- |Acquire a read lock. This will block until there are no pending writers
+-- waiting to acquire the lock.
+acquireRead :: RWLock -> IO ()
```

Comment:
I suppose some of these functions are internal, could they be hidden from external modules?

Reply:
acquireRead, acquireWrite etc.

Reply:
I agree, but the only way to not export this is by listing every other function in the export list which is tedious.